cmd/oras/discover.go

```diff
@@ -136,19 +139,59 @@ func runDiscover(opts discoverOptions) error {
 	return nil
 }
 
-func fetchReferrers(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string) ([]ocispec.Descriptor, error) {
-	results := []ocispec.Descriptor{}
-	err := repo.Referrers(ctx, desc, artifactType, func(referrers []ocispec.Descriptor) error {
-		results = append(results, referrers...)
-		return nil
-	})
-	if err != nil {
-		return nil, err
+func fetchReferrers(ctx context.Context, target oras.ReadOnlyGraphTarget, desc ocispec.Descriptor, artifactType string) ([]ocispec.Descriptor, error) {
+	var results []ocispec.Descriptor
+	if repo, ok := target.(*remote.Repository); ok {
+		// get referrers directly
+		err := repo.Referrers(ctx, desc, artifactType, func(referrers []ocispec.Descriptor) error {
+			results = append(results, referrers...)
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// find matched referrers in all predecessors
+		predecessors, err := target.Predecessors(ctx, desc)
+		if err != nil {
+			return nil, err
+		}
+		for _, node := range predecessors {
+			var fetched []byte
+			if rc, err := target.Fetch(ctx, node); err != nil {
+				return nil, err
+			} else {
```

Comment:
The else enclosing can be deleted.

Reply:
If else is deleted, then `rc` will be not reachable.

Reply:
It reminds me that I can actually use https://pkg.go.dev/oras.land/oras-go/v2#FetchBytes to avoid reading from `rc` myself

Reply:
`rc` is not deferred to be closed.